docs/rfc/rfc-002-zero-copy-encoding.md

```diff
@@ -0,0 +1,402 @@
+# RFC 002: Zero-Copy Encoding
+
+## Changelog
+
+* 2022-03-08: Initial draft
+
+## Background
+
+When the SDK originally migrated to [protobuf state encoding](./../architecture/adr-019-protobuf-state-encoding.md),
+zero-copy encodings such as [Cap'n Proto](https://capnproto.org/)
+and [FlatBuffers](https://google.github.io/flatbuffers/)
+were considered. We considered how a zero-copy encoding could be beneficial for interoperability with modules
+and scripts in other languages and VMs. However, protobuf was still chosen because the maturity of its ecosystem and
+tooling was much higher and the client experience and performance were considered the highest priorities.
+
+In [ADR 033: Protobuf-based Inter-Module Communication](./../architecture/adr-033-protobuf-inter-module-comm.md), the
+idea of cross-language/VM inter-module
+communication was considered again. And in the discussions surrounding [ADR 054: Semver Compatible SDK Modules](./../architecture/adr-054-semver-compatible-modules.md),
+it was determined that multi-language/VM support in the SDK is a near term priority.
+
+While we could do cross-language/VM inter-module communication with protobuf binary or even JSON, the performance
+overhead is deemed to be too high because:
+* we are proposing replacing keeper calls with inter-module message calls and the overhead of even the inter-module
+routing checks has come into question by some SDK users without even considering the possible overhead of encoding.
+Effectively we would be replacing function calls with encoding. One of the SDK's primary objectives currently is
+improving performance, and we want to avoid inter-module calls from becoming a big step backward.
+* we want Rust code to be able to operate in highly resource constrained virtual machines so whatever we can do to
+reduce performance overhead as well as the size of generated code will make it easier and more feasible to deploy
+first-class integrations with these virtual machines.
+
+Thus, the agreement when the [ADR 054](./../architecture/adr-054-semver-compatible-modules.md) working group concluded
+was to pursue a performant zero-copy encoding which is suitable for usage in highly resource constrained environments.
```

Comment:
How do you know that encoding and decoding of these types is too slow? What performance profiles can you show? What benchmarks and tests can be used to measure success, failure, and regressions?

Reply:
@tac0turtle shared some info about proto binary not being viable in certain VM environments. Maybe he can share more concrete details. But we are talking about zkVM environments and my understanding is that these introduce quite a bit of overhead such that any optimization counts.

Reply:
If you define an API in terms of protobuf types, then you can't assume any specific details in the encoding of those types which aren't guaranteed by the protobuf spec. Basically, if a zero-copy parse of a message fails, this can't represent a fatal error, because nothing guarantees the message was encoded according to those zero-copy expectations; the parser has to fall back to the "normal" decode behavior. But there is no way to know that a successful zero-copy parse of a message actually means that the message has been accurately parsed! Different parsers are not necessarily mutually exclusive against the same input data. So, as far as I can tell, this basically doesn't work.

Reply:
Of course we can. If we specify an API using this encoding, users of that API should follow those rules. If they're not following the rules, that's user error. Also this is primarily for inter module message passing and RFC 003 which is based on this specifics startup checks to ensure schemas of different modules are aligned and conformant

Reply:
Protobufs don't provide a mechanism for users to dictate anything about how types are encoded. So you can't "specify an API" that claims to speak protobuf, and also expects anything about the encoded form of its types which is above or beyond what is guaranteed by the protobuf spec. 

If you do have expectations about encoded form, that's fine, but in that case you can't claim that you're speaking protobufs any more.

Reply:
I've already addressed this particular comment of yours. Thanks for the input. I'm going to resolve this thread